### PR TITLE
Automatic CSS Variable fallbacks (Components CSS)

### PR DESCRIPTION
**Pull request by ItsJonQ**
Screen Shot 2020-07-21 at 9 29 42 AM

(Image shows a generated CSS declaration fallback for it's `var()` usage in `RangeControl`)

This is a follow up to:
https://github.com/WordPress/gutenberg/pull/24059

This update adds automatic CSS variable fallback handling for CSS-in-JS generated styles from our `@wordpress/components` package (powered by [Emotion](https://emotion.sh/docs/introduction)).

This is accomplished by combining a couple of parts:

* A custom stylis plugin that can transform CSS declaration output
* A custom Emotion provider to target CSS-in-JS generated styles

### ✨ Stylis Plugin

[Stylis](https://github.com/thysultan/stylis.js) is the light-weight, power, and fast (and awesome) CSS compiler that many CSS-in-JS libraries use (e.g. Styled Components, Emotion, Linaria, etc...). Stylis allows us to add custom middleware (aka. plugins) that can modify output. This pathway is enabled by Emotion with the creation of a "custom cache".

### 👩‍🎤 Emotion Provider (and custom cache)

Don't let the custom cache scare you. This is how Emotion works under-the-hood! We're creating our own cache instance (the brains that coordinates everything for Emotion) and syncing it with the components via a Provider.

This is the same technique I introduced in this PR:
https://github.com/WordPress/gutenberg/pull/23215

(I'll need to update that one once this one merges)

### 🙏 IE Only Please

To optimize for performance, the transformations only occur on browsers that **do not support CSS variables**. The plugin auto-detects this by running a `window.CSS.supports()` check.

This option can be enabled/disabled by passing `skipSupportedBrowsers: false` when setting up the plugin:

```
const createCustomCache = memoize( () => {
	return createEmotionCache( {
		stylisPlugins: [
			stylisPluginCssCustomProperties( { skipSupportedBrowsers: false } ),
		],
	} );
} );
```

This is useful for **tests** (but also for debugging)

### 🍿 StyledProvider

This new `<StyledProvider />` enables us to seamlessly enhance our CSS-in-JS workflow by using additional plugins.

Some of these enhancements include...

* Automatic scope boosting (to ensure consistent + predictable CSS overrides)
* Automatic RTL handling

## How has this been tested?

Tested heavily with a barrage of unit tests. These unit tests throw wild (but valid) combinations of CSS variables, fallbacks, and odd value combinations to ensure that the parser works correctly.

The integration with the `<StyledProvider />` (and generated styles) were tested manually in local Gutenberg (for both the editor and Site Edit).

To test it locally...

* Add `skipSupportedBrowsers: false` to the `stylisPluginCssCustomProperties()` plugin
* Run `npm run dev`
* Add a Cover block with an image
* Inspect element on the `RangeSlider` component
* You should see a color fallback

Screen Shot 2020-07-21 at 9 29 42 AM

## Types of changes

* Created a custom stylis plugin
* Created a Provider that hooks it up
* Integrates Provider with EditPost and EditSite
* Removed manual CSS variable fallbacks (that were recently added)

cc'ing @diegohaz <3 (Just in case you're interested)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [n/a] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/styled-provider/index.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import createEmotionCache from '@emotion/cache';
+import { CacheProvider } from '@emotion/core';
+import memoize from 'memize';
+
+/**
+ * Internal dependencies
+ */
+import { stylisPluginCssCustomProperties } from './plugins';
+
+const createCustomCache = memoize( () => {
```
Comment (ItsJonQ) [inline packages/components/src/styled-provider/index.js:13]: Technique suggested by a core Emotion member:
https://github.com/emotion-js/emotion/issues/760#issuecomment-404353706

**Comment by ItsJonQ** [comment]
Update! I released the plugin code under my personal Github:
https://github.com/ItsJonQ/stylis-plugin-css-variables

The reason is because I want to have it available for personal use.
It's been published on npm as well:
https://www.npmjs.com/package/stylis-plugin-css-variables

If folks prefer, we can totally keep the plugin code within the project (leave this PR as is).
Alternatively, we can use the npm package ✌️ 

Open to thoughts!

File: packages/components/src/styled-provider/index.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/**
+ * External dependencies
+ */
+import createEmotionCache from '@emotion/cache';
+import { CacheProvider } from '@emotion/core';
+import stylisPluginCssVariables from 'stylis-plugin-css-variables';
```
Comment (ItsJonQ) [inline packages/components/src/styled-provider/index.js:6]: Code originally existed in this repo as part of this PR.
It was abstracted to a separate repo/package:
https://github.com/ItsJonQ/stylis-plugin-css-variables

**Comment by ItsJonQ** [comment]
cc'ing @saramarcondes 

Haiiiii!!! Continuing our convo from your other CSS-in-JS PRs 😊 .

I think this StyledProvider will help smoother over CSS Variable fallback issues we talked about. 
The `ui.theme` color's value has a fallback in it's value. That means that it should work for IE11 with my Stylis plugin.

Do you think you can take a quick look to make sure it works and the implementation feels okay?

If so, maybe we can proceed with adding the resolver in there as well (perhaps as a hook within `StyledProvider`)

File: packages/editor/src/components/provider/index.js [hunk 3]
```diff
@@ -271,7 +272,7 @@ class EditorProvider extends Component {
 		);
 
 		return (
-			<>
+			<StyledProvider>
```
Comment (youknowriad) [inline packages/editor/src/components/provider/index.js:275]: This enables it for the post-editor (probably better moved to edit-post) but not for edit-widgets, edit-site, edit-navigation and list-reusabale-blocks (these are the top level packages).

 - How do we make sure folks don't forget to add these providers? 
 - Is there a way to throw a warning if we detect a missing provider?
 - Should we embed other providers there in a single one? (thinking about popover and slot providers).
 - Should this be marked as a breaking change in the changelog of the components package?

**Review by youknowriad — COMMENTED** [review]
Thanks for the PR @ItsJonQ I like the new automatic abstraction.

I don't like "Providers" in general because they make components less easily reusable but it feels that at least a single provider is mandatory to use the WordPress components and I wonder if should rename it to be able to add any addition providers required later to the same one.

**Comment by ItsJonQ** [comment]
> Thanks for the PR @ItsJonQ I like the new automatic abstraction.

@youknowriad 

Wonderful :)

I do not like Providers either, especially if they are mandatory (which sometimes cannot be avoided).

### 🌄 Background

For context, I revisited this PR because @saramarcondes started to graciously help out with components by refactoring styles to use CSS-in-JS. In doing so, she's run into a lot of roadblocks, many of which were things I've previously experienced. 

### 🙈 Roadblocks

The primary roadblock would be converting the styles from Sass -> CSS-in-JS safely without introducing side-effects. This is quite tricky due to how (unexpected) coupled certain styles are.

Well intentioned style rules that targeted deeper HTML elements (therefore raising specificity) that happen organically over time.

That being said, these speed bumps are bound to happen regardless of we're doing a 1:1 Sass -> CSS-in-JS conversion or if we're doing something larger (like rewriting components like in [G2 Components](https://github.com/ItsJonQ/g2/tree/master/packages/components/src)).

### 👉 Next Steps

After some decision with @saramarcondes, we felt like it may be best to pause on some of the CSS-in-JS conversion and focus more on G2 Components (Sara, please correct me if I'm wrong). The reason is because many of the issues Sara and I have encountered were resolved within G2's Style System (if not 100% resolved, then at least something to allow us to strategically iterate). We've also been testing out integration strategies for G2 x Gutenberg.

--

P.S. Back to Providers! In contrast, [G2's Style System](https://github.com/ItsJonQ/g2/tree/master/packages/styles) would not require a Provider to achieve this same effect, as the compiling via plugins (same basic technique) happens at a lower level. The only cause where a provider would be needed would be iFrame based rendering.

**Comment by ItsJonQ** [comment]
Closing this up for now! Hopefully we'll be able get this from [G2 Components](https://github.com/ItsJonQ/g2/blob/master/packages/create-styles/src/createCompiler/plugins/cssVariables.js).

**Pull request closed without merging**